### PR TITLE
ObjC: Test the simple parse helper and fix an edge case with errors in the final line

### PR DESCRIPTION
**Pull request by thomasvl**

### CODE REVIEW

File: src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc [hunk 1]
```diff
@@ -242,6 +243,135 @@ TEST(ObjCHelperDeathTest, TextFormatDecodeData_Failures) {
...
+      EXPECT_TRUE(ParseSimpleStream(input, "dummy", &collector, &err_str));
+      EXPECT_EQ(lines, test.second);
+      EXPECT_TRUE(err_str.empty());
+    }
+  }
+}
+
+TEST(ObjCHelper, ParseSimple_RejectLines) {
+  const std::vector<std::tuple<std::string, std::string, int>> tests = {
+    std::make_tuple("a\nb\nc", "a", 1),
```
Comment (thomasvl) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc:332]: If I just make these values `{"a\nb\nc", "a", 1}`, this it worked on macOS, but failed on some of the [linux builds](https://source.cloud.google.com/results/invocations/f8912da0-b6c7-4065-8476-a31e2ba84631/targets/protobuf%2Fgithub%2Fmaster%2Fubuntu%2Fcpp_distcheck%2Fpresubmit/log).  Any idea if there's a better way to do this initialization?

File: src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc [hunk 2]
```diff
@@ -242,6 +243,135 @@ TEST(ObjCHelperDeathTest, TextFormatDecodeData_Failures) {
...
+    return true;
+  }
+
+ private:
+  std::vector<std::string>* lines_;
+  const std::string* reject_;
+  bool skip_msg_;
+};
+
+const int kBlockSizes[] = {-1, 1, 2, 5, 64};
```
Comment (dmaclach) [inline src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc:274]: Any particular reasons for these blocksizes? I know that -1 is default, but others?

File: src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc (hunk 1 above)
Comment (dmaclach) [reply src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc:332]: hmm.. wonder if it's a const_expr problem. the make_tuple works? If so I would leave it.

**Review by dmaclach — APPROVED** [review]

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc [hunk 3]
```diff
@@ -1638,69 +1638,69 @@ class Parser {
...
-    if (line.empty()) {
-      continue;  // Blank line.
-    }
-    if (!line_consumer_->ConsumeLine(line, &error_str_)) {
-      return false;
-    }
-  }
-  return true;
+std::string FullErrorString(const std::string& name, int line_num, const std::string& msg) {
+  return std::string("error: ") + name + " Line " + StrCat(line_num) + ", " + msg;
```
[the remark this answers is not in the document]
Comment (thomasvl) [reply src/google/protobuf/compiler/objectivec/objectivec_helpers.cc:1703]: `name` ends up being the file name when we read things in.  Most cases, that formatting doesn't actually appear to be working for all the places see it, so just keeping something mostly readable.

File: src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc (hunk 2 above)
Comment (thomasvl) [reply src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc:274]: just some chunks to ensure the line buffering works out correctly.

File: src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc (hunk 1 above)
Comment (thomasvl) [reply src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc:332]: I didn't check, but I guess linux could be gcc or something like that, also adding to the _stl fun_. ☹️  Since this works, I'll just keep it, wish it made more sense to me.

File: src/google/protobuf/compiler/objectivec/objectivec_helpers.cc [hunk 4]
```diff
@@ -1638,69 +1638,69 @@ class Parser {
   Parser(LineConsumer* line_consumer)
       : line_consumer_(line_consumer), line_(0) {}
 
-  // Parses a check of input, returning success/failure.
-  bool ParseChunk(StringPiece chunk);
+  // Feeds in some input, prase what it can, returning success/failure. Calling
```
[the remark this answers is not in the document]
Comment (thomasvl) [reply src/google/protobuf/compiler/objectivec/objectivec_helpers.cc]: fixed

File: src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc [hunk 5]
```diff
@@ -242,6 +243,135 @@ TEST(ObjCHelperDeathTest, TextFormatDecodeData_Failures) {
...
+class TestLineCollector : public LineConsumer {
+ public:
+  TestLineCollector(std::vector<std::string>* inout_lines,
+                    const std::string* reject_line = nullptr,
+                    bool skip_msg = false)
+    : lines_(inout_lines), reject_(reject_line), skip_msg_(skip_msg) {}
+
+  bool ConsumeLine(const StringPiece& line, std::string* out_error) override {
+    if (reject_) {
+      if (*reject_ == line) {
```
[the remark this answers is not in the document]
Comment (thomasvl) [reply src/google/protobuf/compiler/objectivec/objectivec_helpers_unittest.cc]: code evolution strikes again.  fixed.
